pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,297 @@
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+// TelnetConn is the struct representing the telnet connection
+type TelnetConn struct {
+	// conn is the underlying connection
+	conn net.Conn
+	// the connection write channel. Everything required to be written to the connection goes to this channel
+	writeCh           chan []byte
+	srvrOptsLock      sync.Mutex
+	unackedServerOpts map[byte]bool
+	clientOptsLock    sync.Mutex
+	unackedClientOpts map[byte]bool
+	//server            *TelnetServer
+	serverOpts map[byte]bool
+	clientOpts map[byte]bool
+
+	// dataRW is the data buffer. It is written to by the FSM and read from by the data handler
+	dataRW        io.ReadWriter
+	cmdBuffer     bytes.Buffer
+	fsm           *telnetFSM
+	handlerWriter io.WriteCloser
+	// cmdHandler is the command handler for the telnet server
+	// it is a callback function when receiving commands issued by the telnet client
+	// it is wrapped by the cmdHandlerWrapper
+	cmdHandler CmdHandlerFunc
+
+	// dataHandler is the data handler of the telnet server
+	// it is a call back function when receiving data from the telnet client
+	// it is wrapped by the dataHandlerWrapper
+	dataHandler DataHandlerFunc
+
+	// used in the dataHandlerWrapper to notify that the telnet connection is closed
+	dataHandlerCloseCh chan chan struct{}
+	// used in the dataHandlerWrapper to notify that data has been writeen to the dataRW buffer
+	dataWrittenCh chan bool
+
+	// callBack function for receiving an option during negotiation
+	optionCallback func(byte, byte)
+
+	// connWriteDoneCh closes the write loop when the telnet connection is closed
+	connWriteDoneCh chan chan struct{}
+	// negotiationDone notifies that telnet negotiation has been done
+	negotiationDone chan struct{}
+
+	closedMutex sync.Mutex
+	closed      bool
+}
+
+// Safely read/write concurrently to the data Buffer
+// databuffer is written to by the FSM and it is read from by the dataHandler
+type dataReadWriter struct {
+	dataBuffer bytes.Buffer
+	dataMux    *sync.Mutex
+}
+
+func (drw *dataReadWriter) Read(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Read(p)
+}
+
+func (drw *dataReadWriter) Write(p []byte) (int, error) {
+	drw.dataMux.Lock()
+	defer drw.dataMux.Unlock()
+	return drw.dataBuffer.Write(p)
+}
+
+type connectionWriter struct {
+	ch chan []byte
+}
+
+func (cw *connectionWriter) Write(b []byte) (int, error) {
+	if cw.ch != nil {
+		cw.ch <- b
+	}
+	return len(b), nil
+}
+
+func (cw *connectionWriter) Close() error {
+	close(cw.ch)
+	cw.ch = nil
+	return nil
+}
+
+func newTelnetConn(opts connOpts) *TelnetConn {
```

Comment:
@caglar10ur like TCPConn, the only way to get an instance of TelnetConn is through the server's Accept() function. The user of the library should not be able to create a telnet connection in any different way.

For the ConnOpts, it is a small struct and it signals to the user of the library that the library is not going to change the values of the options that are passed in